diff --git a/plutus-core/executables/uplc/Main.hs b/plutus-core/executables/uplc/Main.hs
index 29c4b301bd..b80f96e2c3 100644
--- a/plutus-core/executables/uplc/Main.hs
+++ b/plutus-core/executables/uplc/Main.hs
@@ -209,7 +209,7 @@ runEval (EvalOptions inp ifmt printMode budgetMode traceMode outputMode timingMo
                                 _   -> error "Timing evaluations returned inconsistent results"
                 handleResults t (res, budget, logs) = do
                     case res of
-                        Left err -> hPrint stderr err >> exitFailure
+                        Left err -> hPrint stderr err
                         Right v  -> writeToFileOrStd outputMode (show (getPrintMethod printMode v))
                     case budgetMode of
                         Silent    -> pure ()
@@ -217,6 +217,9 @@ runEval (EvalOptions inp ifmt printMode budgetMode traceMode outputMode timingMo
                     case traceMode of
                         None -> pure ()
                         _    -> writeToFileOrStd outputMode (T.unpack (T.intercalate "\n" logs))
+                    case res of
+                        Left _ -> exitFailure
+                        Right _  -> pure ()
 
 ----------------- Print examples -----------------------
 runUplcPrintExample ::
